# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # Init
    source_id_text = []
    target_id_text = []
       
    print( "Source text = ", source_text )
    print( "Target text = ", target_text )
    
    # Loop over each element in source_text, look up its id in source_vocab_to_int, and add to list source_id_text
    # For some reason, this looping construct won't pass the unit tests!  Wrong number of elements...
#    for s_line in source_text.split('\n'):
#        for s_word in s_line.split():
#           source_id_text.append( source_vocab_to_int.get( s_word ) )

    # Use list comprenhensions instead; assistance from Mihai Lai and David Katz in Slack forums
    source_id_text = [[source_vocab_to_int.get( s_word ) for s_word in s_line.split()] for s_line in source_text.split('\n') ]

#    for t_line in target_text.split('\n'): 
#        for t_word in t_line.split():
#           target_id_text.append( target_vocab_to_int.get( t_word ))
    target_id_text = [[target_vocab_to_int.get( t_word ) for t_word in t_line.split()] + [target_vocab_to_int['<EOS>']] for t_line in target_text.split('\n')]

    # Update last element        
    #target_id_text.append( target_vocab_to_int['<EOS>'] )            

    print( "Source id len = ", len(source_id_text) )
    print( "Target id len = ", len(target_id_text) )
    
    return (source_id_text, target_id_text)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Source text =  new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
Target text =  new jersey est parfois calme pendant l' automne , et il est neigeux en avril .
les états-unis est généralement froid en juillet , et il gèle habituellement en novembre .
california est généralement calme en mars , et il est généralement chaud en juin .
les états-unis est parfois légère en juin , et il fait froid en septembre .
Source id len =  4
Target id len =  4
Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.


Source id len =  137861
Target id len =  137861


# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [5]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [6]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.1.0
Default GPU Device: /gpu:0


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [7]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    input = tf.placeholder( tf.int32, [None, None], name='input' )
    targets = tf.placeholder( tf.int32, [None, None], name='targets' )
    learning_rate = tf.placeholder( tf.float32, name ='learning_rate')
    keep_probability = tf.placeholder( tf.float32, name ='keep_prob') 
    target_sequence_length = tf.placeholder( tf.int32, [None], name='target_sequence_length' )
    max_target_sequence_length = tf.reduce_max( target_sequence_length )
    source_sequence_length =tf.placeholder( tf.int32, [None], name='source_sequence_length' )
    
    return (input, targets, learning_rate, keep_probability, target_sequence_length,
            max_target_sequence_length, source_sequence_length)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [8]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placeholder (tensor)
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # Algorithm:
    # Chunk out target_data in batch_size pieces...
    # Remove last word from each batch
    # add GO ID to the beginning of each batch
    #print("Target data = ", target_data)
    #print("Target data size = ", tf.size( target_data ))
    #new_target_data = []
    
    #go_id = target_vocab_to_int.get( '<GO>' )
    #num_batches = len( target_data ) / batch_size   <== this approach fails because you can't take the length of a tensor like this
    
    # for each batch / num_batches
    #for batch_index in range( 0, num_batches ):
    #   curr_batch = target_data[ batch_index * batch_size : ( batch_index + 1 ) * batch_size - 1 ]
    #   new_batch = curr_batch[1:]      # cut off the first element
    #   new_batch.append( go_id )  # add go-id to end
    
    #  new_target_data.extend( new_batch )
    
    #return new_target_data
    
    # Using snippet from sequence_to_sequence implementation:
    ending = tf.strided_slice(target_data, [0, 0], [batch_size, -1], [1, 1])
    new_target_data = tf.concat([tf.fill([batch_size, 1], target_vocab_to_int['<GO>']), ending], 1)
    
    return new_target_data

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [9]:
from imp import reload
reload(tests)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    # Used snippet from sequence to sequence notebook's function encoding_layer()
    # Embed the encoder input using tf.contrib.layers.embed_sequence
    enc_embedding = tf.contrib.layers.embed_sequence( rnn_inputs, 
                                                      source_vocab_size, 
                                                      encoding_embedding_size )
    
    # Construct a stacked tf.contrib.rnn.LSTMCell wrapped in a tf.contrib.rnn.DropoutWrapper
    def build_cell( r_size ):
        e_cell = tf.contrib.rnn.LSTMCell( rnn_size, 
                                          initializer = tf.random_uniform_initializer( -0.1, 0.1, seed=2))
        return e_cell
    
    encoding_cell = tf.contrib.rnn.MultiRNNCell( [ build_cell( rnn_size ) for _ in range(num_layers )])
    
    drop_cell = tf.contrib.rnn.DropoutWrapper( encoding_cell, 
                                               output_keep_prob=keep_prob)
    
    # Pass cell and embedded input to tf.nn.dynamic_rnn()    
    RNN_output, RNN_state = tf.nn.dynamic_rnn( drop_cell, 
                                               enc_embedding, 
                                               sequence_length=source_sequence_length, 
                                               dtype=tf.float32  )
    
    return RNN_output, RNN_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [10]:

def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    #with tf.variable_scope("decode"):
        # Create a tf.contrib.seq2seq.TrainingHelper
    train_help = tf.contrib.seq2seq.TrainingHelper( inputs          = dec_embed_input,
                                                    sequence_length = target_sequence_length,
                                                    time_major      = False )

    # Create a tf.contrib.seq2seq.BasicDecoder
    base_decode = tf.contrib.seq2seq.BasicDecoder( dec_cell,
                                                   train_help,
                                                   encoder_state,
                                                   output_layer )

    # Obtain the decoder outputs from tf.contrib.seq2seq.dynamic_decode
    decoder_out = tf.contrib.seq2seq.dynamic_decode( base_decode,
                                                    impute_finished = True,
                                                    maximum_iterations = max_summary_length )[0]

    return decoder_out



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [11]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    start_tokens = tf.tile(tf.constant([start_of_sequence_id], dtype=tf.int32), [batch_size], name='start_tokens')

    # Create a tf.contrib.seq2seq.GreedyEmbeddingHelper
    inf_helper = tf.contrib.seq2seq.GreedyEmbeddingHelper( dec_embeddings,
                                                           start_tokens,
                                                           end_of_sequence_id ) # end token
    #print( "geHelper size:", tf.size( geHelper ))

    # Create a tf.contrib.seq2seq.BasicDecoder
    base_decoder = tf.contrib.seq2seq.BasicDecoder( dec_cell,        # cell
                                                    inf_helper,      # helper
                                                    encoder_state,   # init state
                                                    output_layer ) 

    # Obtain the decoder outputs from tf.contrib.seq2seq.dynamic_decode
    inf_decode_output, _ = tf.contrib.seq2seq.dynamic_decode( base_decoder, 
                                                              impute_finished    = True,
                                                              maximum_iterations = max_target_sequence_length)

    return inf_decode_output



"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [12]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
#    Embed the target sequences
    dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)

#    Construct the decoder LSTM cell (just like you constructed the encoder cell above)
    def make_cell( rnn_size ):
        dec_cell = tf.contrib.rnn.LSTMCell( rnn_size ,
                                            initializer=tf.random_uniform_initializer(-0.1, 0.1, seed=2))
        return dec_cell
    
    dec_cell = tf.contrib.rnn.MultiRNNCell([make_cell(rnn_size) for _ in range(num_layers)])

#   Create an output layer to map the outputs of the decoder to the elements of our vocabulary
    output_layer = Dense( target_vocab_size,
                          kernel_initializer = tf.truncated_normal_initializer( mean=0.0, stddev=0.1))

#    Use the your decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob) function to get the training logits.
    with tf.variable_scope("decode"):
        
        train_decoder_out = decoding_layer_train(encoder_state, 
                                                 dec_cell, 
                                                 dec_embed_input, 
                                                 target_sequence_length, 
                                                 max_target_sequence_length, 
                                                 output_layer, 
                                                 keep_prob)

#    Use your decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob) function to get the inference logits.
    with tf.variable_scope("decode", reuse=True):
        
        infer_decoder_out = decoding_layer_infer(encoder_state, 
                                                 dec_cell, 
                                                 dec_embeddings, 
                                                 target_vocab_to_int['<GO>'],  #start_of_sequence_id,
                                                 target_vocab_to_int['<EOS>'], # end_of_sequence_id, 
                                                 max_target_sequence_length,
                                                 target_vocab_size, 
                                                 output_layer, 
                                                 batch_size, 
                                                 keep_prob)
        
    return (train_decoder_out, infer_decoder_out )


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [13]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    # Encode the input using your encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
    # source_sequence_length, source_vocab_size, encoding_embedding_size).
    
    _, encoder_state = encoding_layer(input_data, 
                                       rnn_size, 
                                       num_layers, 
                                       keep_prob, 
                                       source_sequence_length, 
                                       source_vocab_size, 
                                       enc_embedding_size)

    # Process target data using your process_decoder_input(target_data, target_vocab_to_int, batch_size) function.
    dec_input = process_decoder_input(target_data, 
                                      target_vocab_to_int, 
                                      batch_size)
    
    # Decode the encoded input using your decoding_layer(dec_input, enc_state, target_sequence_length, 
    # max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size) function
    train_decoder_out, infer_decoder_out = decoding_layer(dec_input, 
                                                          encoder_state,
                                                          target_sequence_length, 
                                                          max_target_sentence_length,
                                                          rnn_size,
                                                          num_layers, 
                                                          target_vocab_to_int, 
                                                          target_vocab_size,
                                                          batch_size, 
                                                          keep_prob, 
                                                          dec_embedding_size)
    
    
    return (train_decoder_out, infer_decoder_out)


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [18]:
# Number of Epochs
epochs = 50
# Batch Size
batch_size = 128
# RNN Size
rnn_size = 50
# Number of Layers
num_layers = 3
# Embedding Size
encoding_embedding_size = 15
decoding_embedding_size = 15
# Learning Rate
learning_rate = 0.001
# Dropout Keep Probability
keep_probability = 0.5
display_step = 50

### Build the Graph
Build the graph using the neural network you implemented.

In [19]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [20]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [21]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch   50/1077 - Train Accuracy: 0.2590, Validation Accuracy: 0.3356, Loss: 3.6148
Epoch   0 Batch  100/1077 - Train Accuracy: 0.2617, Validation Accuracy: 0.3363, Loss: 3.3077
Epoch   0 Batch  150/1077 - Train Accuracy: 0.3549, Validation Accuracy: 0.3739, Loss: 2.9232
Epoch   0 Batch  200/1077 - Train Accuracy: 0.3559, Validation Accuracy: 0.4116, Loss: 2.8594
Epoch   0 Batch  250/1077 - Train Accuracy: 0.4190, Validation Accuracy: 0.4290, Loss: 2.5530
Epoch   0 Batch  300/1077 - Train Accuracy: 0.3684, Validation Accuracy: 0.4535, Loss: 2.7631
Epoch   0 Batch  350/1077 - Train Accuracy: 0.3980, Validation Accuracy: 0.4656, Loss: 2.6459
Epoch   0 Batch  400/1077 - Train Accuracy: 0.4203, Validation Accuracy: 0.4769, Loss: 2.5149
Epoch   0 Batch  450/1077 - Train Accuracy: 0.4344, Validation Accuracy: 0.4769, Loss: 2.4357
Epoch   0 Batch  500/1077 - Train Accuracy: 0.4562, Validation Accuracy: 0.4922, Loss: 2.3325
Epoch   0 Batch  550/1077 - Train Accuracy: 0.4219, Validati

Epoch   4 Batch  250/1077 - Train Accuracy: 0.6566, Validation Accuracy: 0.6481, Loss: 0.4377
Epoch   4 Batch  300/1077 - Train Accuracy: 0.6813, Validation Accuracy: 0.6328, Loss: 0.4752
Epoch   4 Batch  350/1077 - Train Accuracy: 0.6270, Validation Accuracy: 0.6598, Loss: 0.4744
Epoch   4 Batch  400/1077 - Train Accuracy: 0.6574, Validation Accuracy: 0.6538, Loss: 0.4600
Epoch   4 Batch  450/1077 - Train Accuracy: 0.6711, Validation Accuracy: 0.6403, Loss: 0.4399
Epoch   4 Batch  500/1077 - Train Accuracy: 0.6887, Validation Accuracy: 0.6488, Loss: 0.4376
Epoch   4 Batch  550/1077 - Train Accuracy: 0.6664, Validation Accuracy: 0.6722, Loss: 0.4683
Epoch   4 Batch  600/1077 - Train Accuracy: 0.6815, Validation Accuracy: 0.6701, Loss: 0.4113
Epoch   4 Batch  650/1077 - Train Accuracy: 0.6828, Validation Accuracy: 0.6854, Loss: 0.4336
Epoch   4 Batch  700/1077 - Train Accuracy: 0.6711, Validation Accuracy: 0.6793, Loss: 0.4151
Epoch   4 Batch  750/1077 - Train Accuracy: 0.6910, Validati

Epoch   8 Batch  450/1077 - Train Accuracy: 0.8379, Validation Accuracy: 0.8093, Loss: 0.1659
Epoch   8 Batch  500/1077 - Train Accuracy: 0.8613, Validation Accuracy: 0.7976, Loss: 0.1510
Epoch   8 Batch  550/1077 - Train Accuracy: 0.8082, Validation Accuracy: 0.8136, Loss: 0.1798
Epoch   8 Batch  600/1077 - Train Accuracy: 0.8586, Validation Accuracy: 0.8072, Loss: 0.1540
Epoch   8 Batch  650/1077 - Train Accuracy: 0.8250, Validation Accuracy: 0.8214, Loss: 0.1599
Epoch   8 Batch  700/1077 - Train Accuracy: 0.8637, Validation Accuracy: 0.8207, Loss: 0.1424
Epoch   8 Batch  750/1077 - Train Accuracy: 0.8586, Validation Accuracy: 0.8118, Loss: 0.1564
Epoch   8 Batch  800/1077 - Train Accuracy: 0.8449, Validation Accuracy: 0.8281, Loss: 0.1512
Epoch   8 Batch  850/1077 - Train Accuracy: 0.8281, Validation Accuracy: 0.8125, Loss: 0.1768
Epoch   8 Batch  900/1077 - Train Accuracy: 0.8555, Validation Accuracy: 0.7987, Loss: 0.1702
Epoch   8 Batch  950/1077 - Train Accuracy: 0.8371, Validati

Epoch  12 Batch  650/1077 - Train Accuracy: 0.8875, Validation Accuracy: 0.8864, Loss: 0.0882
Epoch  12 Batch  700/1077 - Train Accuracy: 0.9199, Validation Accuracy: 0.8871, Loss: 0.0621
Epoch  12 Batch  750/1077 - Train Accuracy: 0.9078, Validation Accuracy: 0.8814, Loss: 0.0788
Epoch  12 Batch  800/1077 - Train Accuracy: 0.9117, Validation Accuracy: 0.8913, Loss: 0.0722
Epoch  12 Batch  850/1077 - Train Accuracy: 0.8888, Validation Accuracy: 0.8910, Loss: 0.1001
Epoch  12 Batch  900/1077 - Train Accuracy: 0.9020, Validation Accuracy: 0.8551, Loss: 0.0893
Epoch  12 Batch  950/1077 - Train Accuracy: 0.9156, Validation Accuracy: 0.9052, Loss: 0.0688
Epoch  12 Batch 1000/1077 - Train Accuracy: 0.9182, Validation Accuracy: 0.8640, Loss: 0.0697
Epoch  12 Batch 1050/1077 - Train Accuracy: 0.9273, Validation Accuracy: 0.8800, Loss: 0.0637
Epoch  13 Batch   50/1077 - Train Accuracy: 0.9145, Validation Accuracy: 0.8892, Loss: 0.0651
Epoch  13 Batch  100/1077 - Train Accuracy: 0.9145, Validati

Epoch  16 Batch  850/1077 - Train Accuracy: 0.9055, Validation Accuracy: 0.9357, Loss: 0.0739
Epoch  16 Batch  900/1077 - Train Accuracy: 0.9387, Validation Accuracy: 0.9006, Loss: 0.0674
Epoch  16 Batch  950/1077 - Train Accuracy: 0.9267, Validation Accuracy: 0.9112, Loss: 0.0456
Epoch  16 Batch 1000/1077 - Train Accuracy: 0.9312, Validation Accuracy: 0.9052, Loss: 0.0504
Epoch  16 Batch 1050/1077 - Train Accuracy: 0.9477, Validation Accuracy: 0.9151, Loss: 0.0416
Epoch  17 Batch   50/1077 - Train Accuracy: 0.9336, Validation Accuracy: 0.9112, Loss: 0.0506
Epoch  17 Batch  100/1077 - Train Accuracy: 0.9223, Validation Accuracy: 0.9102, Loss: 0.0479
Epoch  17 Batch  150/1077 - Train Accuracy: 0.9152, Validation Accuracy: 0.9102, Loss: 0.0487
Epoch  17 Batch  200/1077 - Train Accuracy: 0.9102, Validation Accuracy: 0.9091, Loss: 0.0547
Epoch  17 Batch  250/1077 - Train Accuracy: 0.9489, Validation Accuracy: 0.9080, Loss: 0.0453
Epoch  17 Batch  300/1077 - Train Accuracy: 0.9535, Validati

Epoch  20 Batch 1050/1077 - Train Accuracy: 0.9590, Validation Accuracy: 0.9219, Loss: 0.0339
Epoch  21 Batch   50/1077 - Train Accuracy: 0.9477, Validation Accuracy: 0.9368, Loss: 0.0361
Epoch  21 Batch  100/1077 - Train Accuracy: 0.9352, Validation Accuracy: 0.9290, Loss: 0.0347
Epoch  21 Batch  150/1077 - Train Accuracy: 0.9330, Validation Accuracy: 0.9276, Loss: 0.0358
Epoch  21 Batch  200/1077 - Train Accuracy: 0.9262, Validation Accuracy: 0.9311, Loss: 0.0406
Epoch  21 Batch  250/1077 - Train Accuracy: 0.9432, Validation Accuracy: 0.9240, Loss: 0.0369
Epoch  21 Batch  300/1077 - Train Accuracy: 0.9478, Validation Accuracy: 0.9421, Loss: 0.0322
Epoch  21 Batch  350/1077 - Train Accuracy: 0.9445, Validation Accuracy: 0.9396, Loss: 0.0343
Epoch  21 Batch  400/1077 - Train Accuracy: 0.9316, Validation Accuracy: 0.9368, Loss: 0.0439
Epoch  21 Batch  450/1077 - Train Accuracy: 0.9242, Validation Accuracy: 0.9304, Loss: 0.0402
Epoch  21 Batch  500/1077 - Train Accuracy: 0.9305, Validati

Epoch  25 Batch  200/1077 - Train Accuracy: 0.9621, Validation Accuracy: 0.9361, Loss: 0.0308
Epoch  25 Batch  250/1077 - Train Accuracy: 0.9418, Validation Accuracy: 0.9347, Loss: 0.0300
Epoch  25 Batch  300/1077 - Train Accuracy: 0.9511, Validation Accuracy: 0.9407, Loss: 0.0255
Epoch  25 Batch  350/1077 - Train Accuracy: 0.9551, Validation Accuracy: 0.9400, Loss: 0.0278
Epoch  25 Batch  400/1077 - Train Accuracy: 0.9508, Validation Accuracy: 0.9457, Loss: 0.0329
Epoch  25 Batch  450/1077 - Train Accuracy: 0.9316, Validation Accuracy: 0.9283, Loss: 0.0323
Epoch  25 Batch  500/1077 - Train Accuracy: 0.9410, Validation Accuracy: 0.9418, Loss: 0.0252
Epoch  25 Batch  550/1077 - Train Accuracy: 0.9363, Validation Accuracy: 0.9464, Loss: 0.0275
Epoch  25 Batch  600/1077 - Train Accuracy: 0.9394, Validation Accuracy: 0.9350, Loss: 0.0317
Epoch  25 Batch  650/1077 - Train Accuracy: 0.9637, Validation Accuracy: 0.9403, Loss: 0.0253
Epoch  25 Batch  700/1077 - Train Accuracy: 0.9426, Validati

Epoch  29 Batch  400/1077 - Train Accuracy: 0.9563, Validation Accuracy: 0.9503, Loss: 0.0256
Epoch  29 Batch  450/1077 - Train Accuracy: 0.9434, Validation Accuracy: 0.9400, Loss: 0.0266
Epoch  29 Batch  500/1077 - Train Accuracy: 0.9398, Validation Accuracy: 0.9389, Loss: 0.0202
Epoch  29 Batch  550/1077 - Train Accuracy: 0.9359, Validation Accuracy: 0.9531, Loss: 0.0226
Epoch  29 Batch  600/1077 - Train Accuracy: 0.9609, Validation Accuracy: 0.9528, Loss: 0.0235
Epoch  29 Batch  650/1077 - Train Accuracy: 0.9629, Validation Accuracy: 0.9418, Loss: 0.0201
Epoch  29 Batch  700/1077 - Train Accuracy: 0.9652, Validation Accuracy: 0.9506, Loss: 0.0184
Epoch  29 Batch  750/1077 - Train Accuracy: 0.9320, Validation Accuracy: 0.9517, Loss: 0.0249
Epoch  29 Batch  800/1077 - Train Accuracy: 0.9574, Validation Accuracy: 0.9510, Loss: 0.0209
Epoch  29 Batch  850/1077 - Train Accuracy: 0.9479, Validation Accuracy: 0.9538, Loss: 0.0367
Epoch  29 Batch  900/1077 - Train Accuracy: 0.9734, Validati

Epoch  33 Batch  600/1077 - Train Accuracy: 0.9624, Validation Accuracy: 0.9510, Loss: 0.0218
Epoch  33 Batch  650/1077 - Train Accuracy: 0.9688, Validation Accuracy: 0.9577, Loss: 0.0172
Epoch  33 Batch  700/1077 - Train Accuracy: 0.9781, Validation Accuracy: 0.9432, Loss: 0.0160
Epoch  33 Batch  750/1077 - Train Accuracy: 0.9387, Validation Accuracy: 0.9535, Loss: 0.0228
Epoch  33 Batch  800/1077 - Train Accuracy: 0.9816, Validation Accuracy: 0.9659, Loss: 0.0150
Epoch  33 Batch  850/1077 - Train Accuracy: 0.9509, Validation Accuracy: 0.9506, Loss: 0.0342
Epoch  33 Batch  900/1077 - Train Accuracy: 0.9703, Validation Accuracy: 0.9670, Loss: 0.0248
Epoch  33 Batch  950/1077 - Train Accuracy: 0.9643, Validation Accuracy: 0.9496, Loss: 0.0293
Epoch  33 Batch 1000/1077 - Train Accuracy: 0.9557, Validation Accuracy: 0.9393, Loss: 0.0250
Epoch  33 Batch 1050/1077 - Train Accuracy: 0.9711, Validation Accuracy: 0.9478, Loss: 0.0143
Epoch  34 Batch   50/1077 - Train Accuracy: 0.9730, Validati

Epoch  37 Batch  800/1077 - Train Accuracy: 0.9750, Validation Accuracy: 0.9670, Loss: 0.0152
Epoch  37 Batch  850/1077 - Train Accuracy: 0.9535, Validation Accuracy: 0.9606, Loss: 0.0308
Epoch  37 Batch  900/1077 - Train Accuracy: 0.9770, Validation Accuracy: 0.9627, Loss: 0.0254
Epoch  37 Batch  950/1077 - Train Accuracy: 0.9743, Validation Accuracy: 0.9510, Loss: 0.0192
Epoch  37 Batch 1000/1077 - Train Accuracy: 0.9587, Validation Accuracy: 0.9503, Loss: 0.0245
Epoch  37 Batch 1050/1077 - Train Accuracy: 0.9668, Validation Accuracy: 0.9599, Loss: 0.0139
Epoch  38 Batch   50/1077 - Train Accuracy: 0.9820, Validation Accuracy: 0.9599, Loss: 0.0153
Epoch  38 Batch  100/1077 - Train Accuracy: 0.9719, Validation Accuracy: 0.9620, Loss: 0.0160
Epoch  38 Batch  150/1077 - Train Accuracy: 0.9818, Validation Accuracy: 0.9645, Loss: 0.0150
Epoch  38 Batch  200/1077 - Train Accuracy: 0.9809, Validation Accuracy: 0.9602, Loss: 0.0138
Epoch  38 Batch  250/1077 - Train Accuracy: 0.9599, Validati

Epoch  41 Batch 1000/1077 - Train Accuracy: 0.9691, Validation Accuracy: 0.9716, Loss: 0.0160
Epoch  41 Batch 1050/1077 - Train Accuracy: 0.9922, Validation Accuracy: 0.9695, Loss: 0.0106
Epoch  42 Batch   50/1077 - Train Accuracy: 0.9750, Validation Accuracy: 0.9563, Loss: 0.0176
Epoch  42 Batch  100/1077 - Train Accuracy: 0.9781, Validation Accuracy: 0.9624, Loss: 0.0142
Epoch  42 Batch  150/1077 - Train Accuracy: 0.9706, Validation Accuracy: 0.9510, Loss: 0.0159
Epoch  42 Batch  200/1077 - Train Accuracy: 0.9836, Validation Accuracy: 0.9549, Loss: 0.0115
Epoch  42 Batch  250/1077 - Train Accuracy: 0.9670, Validation Accuracy: 0.9727, Loss: 0.0165
Epoch  42 Batch  300/1077 - Train Accuracy: 0.9544, Validation Accuracy: 0.9581, Loss: 0.0177
Epoch  42 Batch  350/1077 - Train Accuracy: 0.9816, Validation Accuracy: 0.9688, Loss: 0.0141
Epoch  42 Batch  400/1077 - Train Accuracy: 0.9793, Validation Accuracy: 0.9620, Loss: 0.0124
Epoch  42 Batch  450/1077 - Train Accuracy: 0.9680, Validati

Epoch  46 Batch  150/1077 - Train Accuracy: 0.9825, Validation Accuracy: 0.9606, Loss: 0.0131
Epoch  46 Batch  200/1077 - Train Accuracy: 0.9879, Validation Accuracy: 0.9663, Loss: 0.0101
Epoch  46 Batch  250/1077 - Train Accuracy: 0.9698, Validation Accuracy: 0.9719, Loss: 0.0159
Epoch  46 Batch  300/1077 - Train Accuracy: 0.9634, Validation Accuracy: 0.9698, Loss: 0.0142
Epoch  46 Batch  350/1077 - Train Accuracy: 0.9738, Validation Accuracy: 0.9638, Loss: 0.0117
Epoch  46 Batch  400/1077 - Train Accuracy: 0.9859, Validation Accuracy: 0.9627, Loss: 0.0129
Epoch  46 Batch  450/1077 - Train Accuracy: 0.9648, Validation Accuracy: 0.9588, Loss: 0.0158
Epoch  46 Batch  500/1077 - Train Accuracy: 0.9812, Validation Accuracy: 0.9691, Loss: 0.0100
Epoch  46 Batch  550/1077 - Train Accuracy: 0.9680, Validation Accuracy: 0.9656, Loss: 0.0112
Epoch  46 Batch  600/1077 - Train Accuracy: 0.9922, Validation Accuracy: 0.9769, Loss: 0.0132
Epoch  46 Batch  650/1077 - Train Accuracy: 0.9816, Validati

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [22]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [23]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [26]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    low_sentence = sentence.lower()
    return [vocab_to_int.get( word, vocab_to_int['<UNK>'] ) for word in low_sentence.split() ]


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [27]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [101, 80, 206, 176, 7, 57, 100]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [142, 98, 346, 326, 326, 105, 223, 204, 1]
  French Words: pourquoi bien au petite petite voiture jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.